File: jazzy_bridge/src/gap_follow/src/reactive_node.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>

#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"

using std::placeholders::_1;

class ReactiveFollowGap : public rclcpp::Node {

   public:
    ReactiveFollowGap() : Node("reactive_node") {
        publisher_ = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>(drive_topic, 10);
        subscription_ = this->create_subscription<sensor_msgs::msg::LaserScan>(lidarscan_topic, 10, std::bind(&ReactiveFollowGap::lidar_callback, this, _1));
    }

   private:
    std::string lidarscan_topic = "/scan";
    std::string drive_topic = "/drive";

    std::vector<double> processed_lidar;

    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr publisher_;
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr subscription_;

    void preprocess_lidar(const sensor_msgs::msg::LaserScan::ConstSharedPtr scan_msg) {
        processed_lidar.clear();
        int window_size = 5;  // smoothing window

        for (size_t i = 0; i < scan_msg->ranges.size(); i++) {
            double avg = 0.0;
            int count = 0;

            for (int offset = -window_size / 2; offset <= window_size / 2; offset++) {
                int idx = i + offset;
                if (idx >= 0 && idx < (int)scan_msg->ranges.size()) {
                    avg += std::min<float>(scan_msg->ranges[idx], 3.0);
                    count++;
                }
            }

            processed_lidar.push_back(avg / count);
        }
    }

    void create_bubble(int closest_point, int bubble_radius) {
        int start_idx = std::max(closest_point - bubble_radius, 0);
        int end_idx = std::min(closest_point + bubble_radius, (int)processed_lidar.size() - 1);
        for (int i = start_idx; i <= end_idx; i++) {
            processed_lidar[i] = 0;
        }
    }

    std::pair<int, int> find_max_gap() {
        int largest_starting_i = 0;
        int longest_gap = 0;
        int curr_gap = 0;
        int curr_start = 0;

        for (size_t i = 0; i < processed_lidar.size(); i++) {
            if (processed_lidar[i] <= 0.5) {
                curr_gap = 0;
                curr_start = i + 1;
            } else {
                curr_gap++;
                if (curr_gap > longest_gap) {
                    largest_starting_i = curr_start;
                    longest_gap = curr_gap;
                }
            }
        }
        return {largest_starting_i, longest_gap};
    }

    int find_best_point(int starting_i, int gap_distance) {
        return starting_i + gap_distance / 2;  // simple midpoint for stability
    }

    double to_radians(double theta) {
        return M_PI * theta / 180.0;
    }

    void lidar_callback(const sensor_msgs::msg::LaserScan::ConstSharedPtr scan_msg) {
        preprocess_lidar(scan_msg);

        // Find closest point to LiDAR
        auto closest_point_iter = std::min_element(processed_lidar.begin(), processed_lidar.end());
        int closest_point = std::distance(processed_lidar.begin(), closest_point_iter);

        // Eliminate all points inside 'bubble'
        create_bubble(closest_point, 30);

        // Find max length gap
        auto max_gap = find_max_gap();

        // Find the best point in the gap
        int best_angle_i = find_best_point(max_gap.first, max_gap.second);

        // Publish Drive message
        auto drive_msg = ackermann_msgs::msg::AckermannDriveStamped();
        drive_msg.drive.steering_angle = scan_msg->angle_min + best_angle_i * scan_msg->angle_increment;
        RCLCPP_INFO(this->get_logger(), "Steering angle: '%f'", drive_msg.drive.steering_angle);

        if (std::abs(drive_msg.drive.steering_angle) < to_radians(10)) {
            drive_msg.drive.speed = 1.5;
        } else if (std::abs(drive_msg.drive.steering_angle) < to_radians(20)) {
            drive_msg.drive.speed = 1.0;
        } else {
            drive_msg.drive.speed = 0.5;
        }
        publisher_->publish(drive_msg);
    }
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ReactiveFollowGap>());
    rclcpp::shutdown();
    return 0;
}

File: jazzy_bridge/src/gap_follow/scripts/reactive_node.py
import rclpy
from rclpy.node import Node

import numpy as np
from sensor_msgs.msg import LaserScan
from ackermann_msgs.msg import AckermannDriveStamped, AckermannDrive

class ReactiveFollowGap(Node):
    """ 
    Implement Wall Following on the car
    """
    def __init__(self):
        super().__init__('reactive_node')
        # Topics & Subs, Pubs
        lidarscan_topic = '/scan'
        drive_topic = '/drive'

        self.scan_subscriber = self.create_subscription(
            LaserScan, lidarscan_topic, self.lidar_callback, 10)
        self.drive_publisher = self.create_publisher(
            AckermannDriveStamped, drive_topic, 10)

    def preprocess_lidar(self, ranges):
        proc_ranges = np.array(ranges)
        proc_ranges[proc_ranges > 3.0] = 3.0
        proc_ranges = np.convolve(proc_ranges, np.ones(5)/5, 'same')
        return proc_ranges

    def find_max_gap(self, free_space_ranges):
        masked = np.ma.masked_equal(free_space_ranges, 0)
        slices = np.ma.clump_unmasked(masked)
        max_gap = max(slices, key=lambda s: s.stop - s.start)
        return max_gap.start, max_gap.stop

    def find_best_point(self, start_i, end_i, ranges):
        return np.argmax(ranges[start_i:end_i]) + start_i

    def lidar_callback(self, data):
        ranges = np.array(data.ranges)
        proc_ranges = self.preprocess_lidar(ranges)

        # Find closest point to LiDAR
        closest_point = proc_ranges.argmin()

        # Eliminate all points inside 'bubble' (set them to zero)
        bubble_radius = 50
        min_idx = max(closest_point - bubble_radius, 0)
        max_idx = min(closest_point + bubble_radius, len(proc_ranges) - 1)
        proc_ranges[min_idx:max_idx] = 0

        # Find max length gap
        gap_start, gap_end = self.find_max_gap(proc_ranges)

        # Find the best point in the gap
        best_point = self.find_best_point(gap_start, gap_end, proc_ranges)

        # Calculate steering angle based on best point
        angle = (best_point - len(proc_ranges)/2) * data.angle_increment

        # Publish Drive message
        drive_msg = AckermannDriveStamped()
        drive_msg.drive.steering_angle = angle
        drive_msg.drive.speed = 1.0

        self.drive_publisher.publish(drive_msg)

def main(args=None):
    rclpy.init(args=args)
    print("WallFollow Initialized")
    reactive_node = ReactiveFollowGap()
    rclpy.spin(reactive_node)

    reactive_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
